snark/src/r1cs.rs

```diff
@@ -0,0 +1,70 @@
+use ark_relations::r1cs::{R1CS, ConstraintSynthesizer};
```

Comment:
These are specialized APIs for working with the existing `ConstraintSystem`-based API

Reply:
Instead of having these segregated APIs, another idea would be to add a `RelationProvider` trait that is a generalization of `RelationProvider`.

```rust
pub trait RelationProvider<R: Relation> {
	// These are all optional because they might not exist always.
	// For example, in indexing phase, there is no assignment,
	// while during proving there may be no index.
	fn index(&self) -> Option<R::Index>; // Maybe this should be `Rc<RefCell<R::Index>>`? to avoid copying.
	fn instance(&self) -> Option<R::Instance>;
	fn witness(&self) -> Option<R::Witness>;
}

pub trait SNARK<R: Relation> {
	fn prove<RP: RelationProvider<R>, ...>(pk, rp: RP)
	// similar for other methods.
}

impl<F: Field> RelationProvider<R1CS<F>> for ConstraintSynthesizer<F> {
// do stuff.
}
```